File: src/systems/GameState.ts
import type { TileKind } from '../world/Grid';

const SAVE_KEY = 'pocket-aquarium:save:v1';

export interface PlacedTank {
  id: string;
  col: number;
  row: number;
  w: number;
  h: number;
  fishSpeciesIds: string[];
}

export interface PlacedDecor {
  id: string;
  col: number;
  row: number;
  kind: 'plant' | 'rock';
}

export interface PlacedPath {
  col: number;
  row: number;
}

export interface SaveData {
  version: 1;
  money: number;
  reputation: number;
  tanks: PlacedTank[];
  decor: PlacedDecor[];
  paths: PlacedPath[];
  caught: string[];
}

export type BuildTool =
  | { kind: 'none' }
  | { kind: 'path' }
  | { kind: 'tank'; size: { w: number; h: number } }
  | { kind: 'decor'; decor: 'plant' | 'rock' }
  | { kind: 'fish'; speciesId: string }
  | { kind: 'erase' };

export interface GameStateData {
  money: number;
  reputation: number;
  tanks: PlacedTank[];
  decor: PlacedDecor[];
  paths: PlacedPath[];
  caught: Set<string>;
  tool: BuildTool;
}

type Listener = (state: GameStateData) => void;

export class GameState {
  private data: GameStateData;
  private listeners = new Set<Listener>();

  constructor() {
    this.data = this.load() ?? this.defaultState();
  }

  get snapshot(): GameStateData {
    return this.data;
  }

  subscribe(fn: Listener): () => void {
    this.listeners.add(fn);
    fn(this.data);
    return () => this.listeners.delete(fn);
  }

  setTool(tool: BuildTool): void {
    this.data.tool = tool;
    this.emit();
  }

  spend(amount: number): boolean {
    if (this.data.money < amount) return false;
    this.data.money -= amount;
    this.emit();
    return true;
  }

  earn(amount: number): void {
    this.data.money += amount;
    this.emit();
  }

  addReputation(amount: number): void {
    this.data.reputation = Math.max(0, this.data.reputation + amount);
    this.emit();
  }

  recordCatch(speciesId: string): void {
    if (!this.data.caught.has(speciesId)) {
      this.data.caught.add(speciesId);
      this.emit();
    }
  }

  addTank(tank: PlacedTank): void {
    this.data.tanks.push(tank);
    this.emit();
  }

  addDecor(decor: PlacedDecor): void {
    this.data.decor.push(decor);
    this.emit();
  }

  addPath(path: PlacedPath): void {
    this.data.paths.push(path);
    this.emit();
  }

  removeAt(col: number, row: number): TileKind | null {
    const tankIdx = this.data.tanks.findIndex(
      (t) => col >= t.col && col < t.col + t.w && row >= t.row && row < t.row + t.h,
    );
    if (tankIdx >= 0) {
      this.data.tanks.splice(tankIdx, 1);
      this.emit();
      return 'tank';
    }
    const decorIdx = this.data.decor.findIndex((d) => d.col === col && d.row === row);
    if (decorIdx >= 0) {
      this.data.decor.splice(decorIdx, 1);
      this.emit();
      return 'decor';
    }
    const pathIdx = this.data.paths.findIndex((p) => p.col === col && p.row === row);
    if (pathIdx >= 0) {
      this.data.paths.splice(pathIdx, 1);
      this.emit();
      return 'path';
    }
    return null;
  }

  findTankAt(col: number, row: number): PlacedTank | null {
    return (
      this.data.tanks.find(
        (t) => col >= t.col && col < t.col + t.w && row >= t.row && row < t.row + t.h,
      ) ?? null
    );
  }

  save(): void {
    const serial: SaveData = {
      version: 1,
      money: this.data.money,
      reputation: this.data.reputation,
      tanks: this.data.tanks,
      decor: this.data.decor,
      paths: this.data.paths,
      caught: [...this.data.caught],
    };
    try {
      localStorage.setItem(SAVE_KEY, JSON.stringify(serial));
    } catch {
      /* storage unavailable — silent */
    }
  }

  reset(): void {
    this.data = this.defaultState();
    try {
      localStorage.removeItem(SAVE_KEY);
    } catch {
      /* noop */
    }
    this.emit();
  }

  private load(): GameStateData | null {
    try {
      const raw = localStorage.getItem(SAVE_KEY);
      if (!raw) return null;
      const parsed = JSON.parse(raw) as SaveData;
      if (parsed.version !== 1) return null;
      return {
        money: parsed.money,
        reputation: parsed.reputation,
        tanks: parsed.tanks,
        decor: parsed.decor,
        paths: parsed.paths,
        caught: new Set(parsed.caught),
        tool: { kind: 'none' },
      };
    } catch {
      return null;
    }
  }

  private defaultState(): GameStateData {
    return {
      money: 500,
      reputation: 0,
      tanks: [],
      decor: [],
      paths: [],
      caught: new Set(),
      tool: { kind: 'none' },
    };
  }

  private emit(): void {
    for (const fn of this.listeners) fn(this.data);
  }
}

export const gameState = new GameState();

File: PocketAquarium.Godot/scripts/systems/GameState.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json.Serialization;
using PocketAquarium;

namespace PocketAquarium.Systems;

public sealed record PlacedTank(
    [property: JsonPropertyName("id")] string Id,
    [property: JsonPropertyName("col")] int Col,
    [property: JsonPropertyName("row")] int Row,
    [property: JsonPropertyName("w")] int W,
    [property: JsonPropertyName("h")] int H,
    [property: JsonPropertyName("fishSpeciesIds")] IReadOnlyList<string> FishSpeciesIds
);

public sealed record PlacedDecor(
    [property: JsonPropertyName("id")] string Id,
    [property: JsonPropertyName("col")] int Col,
    [property: JsonPropertyName("row")] int Row,
    [property: JsonPropertyName("kind")] DecorKind Kind
);

public sealed record PlacedPath(
    [property: JsonPropertyName("col")] int Col,
    [property: JsonPropertyName("row")] int Row
);

public abstract record BuildTool
{
    public sealed record None() : BuildTool;
    public sealed record Path() : BuildTool;
    public sealed record Tank(int W, int H) : BuildTool;
    public sealed record Decor(DecorKind Kind) : BuildTool;
    public sealed record Fish(string SpeciesId) : BuildTool;
    public sealed record Erase() : BuildTool;
}

public sealed class GameState
{
    private readonly List<PlacedTank> _tanks = new();
    private readonly List<PlacedDecor> _decor = new();
    private readonly List<PlacedPath> _paths = new();
    private readonly HashSet<string> _caught = new(StringComparer.Ordinal);

    public int Money { get; private set; }
    public int Reputation { get; private set; }
    public int TicketPrice { get; private set; }
    public IReadOnlyList<PlacedTank> Tanks => _tanks;
    public IReadOnlyList<PlacedDecor> Decor => _decor;
    public IReadOnlyList<PlacedPath> Paths => _paths;
    public IReadOnlySet<string> Caught => _caught;
    public BuildTool Tool { get; private set; } = new BuildTool.None();

    public event Action<GameState>? Changed;

    public GameState()
    {
        ApplyState(
            money: 500,
            reputation: 0,
            ticketPrice: GameConstants.DefaultTicketPrice,
            tanks: Array.Empty<PlacedTank>(),
            decor: Array.Empty<PlacedDecor>(),
            paths: Array.Empty<PlacedPath>(),
            caught: Array.Empty<string>(),
            emit: false
        );
    }

    internal GameState(
        int money,
        int reputation,
        int ticketPrice,
        IEnumerable<PlacedTank> tanks,
        IEnumerable<PlacedDecor> decor,
        IEnumerable<PlacedPath> paths,
        IEnumerable<string> caught
    )
    {
        ApplyState(money, reputation, ticketPrice, tanks, decor, paths, caught, emit: false);
    }

    public bool Spend(int amount)
    {
        if (Money < amount)
            return false;

        Money -= amount;
        Emit();
        return true;
    }

    public void Earn(int amount)
    {
        Money += amount;
        Emit();
    }

    public void AddReputation(int amt)
    {
        Reputation = Math.Max(0, Reputation + amt);
        Emit();
    }

    public void RecordCatch(string speciesId)
    {
        if (_caught.Add(speciesId))
            Emit();
    }

    public void AddTank(PlacedTank t)
    {
        _tanks.Add(t);
        Emit();
    }

    public void AddDecor(PlacedDecor d)
    {
        _decor.Add(d);
        Emit();
    }

    public void AddPath(PlacedPath p)
    {
        _paths.Add(p);
        Emit();
    }

    public string? RemoveAt(int col, int row)
    {
        var tankIdx = _tanks.FindIndex(t => Contains(t, col, row));
        if (tankIdx >= 0)
        {
            _tanks.RemoveAt(tankIdx);
            Emit();
            return "tank";
        }

        var decorIdx = _decor.FindIndex(d => d.Col == col && d.Row == row);
        if (decorIdx >= 0)
        {
            _decor.RemoveAt(decorIdx);
            Emit();
            return "decor";
        }

        var pathIdx = _paths.FindIndex(p => p.Col == col && p.Row == row);
        if (pathIdx >= 0)
        {
            _paths.RemoveAt(pathIdx);
            Emit();
            return "path";
        }

        return null;
    }

    public PlacedTank? FindTankAt(int col, int row) =>
        _tanks.FirstOrDefault(t => Contains(t, col, row));

    public void SetTool(BuildTool t)
    {
        Tool = t;
        Emit();
    }

    public void AdjustTicketPrice(int delta) => SetTicketPrice(TicketPrice + delta);

    public void SetTicketPrice(int price)
    {
        TicketPrice = Math.Clamp(price, GameConstants.MinTicketPrice, GameConstants.MaxTicketPrice);
        Emit();
    }

    public void Reset()
    {
        ApplyState(
            money: 500,
            reputation: 0,
            ticketPrice: GameConstants.DefaultTicketPrice,
            tanks: Array.Empty<PlacedTank>(),
            decor: Array.Empty<PlacedDecor>(),
            paths: Array.Empty<PlacedPath>(),
            caught: Array.Empty<string>(),
            emit: true
        );
    }

    private static bool Contains(PlacedTank tank, int col, int row) =>
        col >= tank.Col && col < tank.Col + tank.W && row >= tank.Row && row < tank.Row + tank.H;

    private void ApplyState(
        int money,
        int reputation,
        int ticketPrice,
        IEnumerable<PlacedTank> tanks,
        IEnumerable<PlacedDecor> decor,
        IEnumerable<PlacedPath> paths,
        IEnumerable<string> caught,
        bool emit
    )
    {
        Money = money;
        Reputation = Math.Max(0, reputation);
        TicketPrice = Math.Clamp(ticketPrice, GameConstants.MinTicketPrice, GameConstants.MaxTicketPrice);
        Tool = new BuildTool.None();

        _tanks.Clear();
        _tanks.AddRange(tanks);

        _decor.Clear();
        _decor.AddRange(decor);

        _paths.Clear();
        _paths.AddRange(paths);

        _caught.Clear();
        foreach (var speciesId in caught)
            _caught.Add(speciesId);

        if (emit)
            Emit();
    }

    private void Emit() => Changed?.Invoke(this);
}
